DSpace/dspace-replicate
# Update BagIt Packaging

Changes for: https://jira.lyrasis.org/browse/DS-4447

* Adds bagit-support library to handle Bagging and BagProfile
* Create BagItAipWriter to keep track of state when packaging Bags
* Creates two helper classes to help encapsulate information
  * BagBitstream stores the bitstream, parent bundle name, and xml elements for the bitstream
  * XmlElement stores attributes and a body for an xml writer
* Updated packers (Item, Collection, Community, Catalog) to be scoped around collecting data to be written to the Bag
* Drops support for fetch.txt in BagIt packaging

-----------------------------------------
JUnit tests can be viewed and run from #48 

-----------------------------------------
There is some logic which has been shuffled around which isn't covered by the changeset that I want to go over as well:

The `Bag` class has methods and sub classes for creating and writing files which were previously used in each of the `Packer`s 
* Creating and resolving payload files (files within the data directory)
* Bag.FlatWriter for writing properties type files (just writes a String or a key/value pair)
* Bag.XmlWriter for writing the various xml files for bags

This functionality was moved into the `BagItAipWriter`
* Any directories which need to exist are created in both the `packageAip` and `writeXmlMetadata` methods. Path resolution is handled in `packageAip`.
  * The `BagWriter` from bag-support also tries to create a directory for the bag on instantiation
* for writing "flat" files, the `packageAip` method handles creating the file and writes directly to the output stream
* `writeXmlMetadata` is a substitute for the `Bag.XmlWriter`

------------------------------------------
Testing:
1. Create a new DSpace repository with the RTS enabled as an addon
2. Create a Community, Collection, and Item in the repository
   1. Optionally - add a logo, or other information for the given object
3. Under the curation context for the Community, select the `Transfer AIP to storage` task
4. Review the 3 AIPs which were written to the repstore directory of your dspace installation
   1. Each AIP should be in a zip format
   2. Each AIP should have all present members (object.properties, metadata.xml, bitstreams)
